nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/dependencies/DependencyResolution.java

```diff
@@ -178,11 +178,11 @@ public synchronized void sort(PackageManager pm) {
             String id = pkgName + "-" + allPackages.get(pkgName).toString();
             DownloadablePackage pkg = pm.findPackageById(id);
             List<Version> installedVersions = pm.findLocalPackageInstalledVersions(pkg.getName());
-            if (pkg.getPackageState().isInstalled()) {
+            if (pkg.getPackageState().isInstalled() && !orderedRemovablePackages.contains(id)) {
                 // Already installed in the wanted version, nothing to do
                 localUnchangedPackages.put(pkg.getName(), pkg.getVersion());
             } else {
-                if (installedVersions.size() > 0) {
+                if (installedVersions.size() > 0 && !installedVersions.contains(pkg.getVersion())) {
```

Comment:
What is the point and link of those two changes (`!orderedRemovablePackages.contains(id)`, `!installedVersions.contains(pkg.getVersion())`)?
The conditional matrix became a little bit obscure, it would deserve two additional comments :)
```
 if (pkg.getPackageState().isInstalled() && !orderedRemovablePackages.contains(id)) {
                // Already installed in the wanted version, nothing to do
            } else {
                if (installedVersions.size() > 0 && !installedVersions.contains(pkg.getVersion())) {
                    // Upgrade case: already installed in other version(s)
                }
                if (pkg.getPackageState() == PackageState.REMOTE) {
                    // ?
                } else {
                    // ?
                }
}
```